### DISCUSSION TITLE
[Proposal] Target type inference for new() keyword in switch expressions

### DISCUSSION THREAD

**Discussion by I-Omnibus** [dotnet/roslyn#49821] [Ideas]
[ https://dotnetfiddle.net/Widget/CKXaCr ]
```csharp
/// <summary>
/// Target type inference for new() keyword in switch expressions
/// </summary>
using System.Collections.Generic;
using System.Linq;

class CSharpLang {

	void NewKeywordInference() {
		PleaseMakeInference inferred;
		// Target type has been inferred with new() keyword
		inferred = new() { Direction = Move.Nowhere };

		// No inference available here despite the compiler knowing 
		// that it's explicitly dealing IEnumerable<PleaseMakeInference>
		var values = new int[] { -2, -1, 0, 1, 2 };
		IEnumerable<PleaseMakeInference> directions = values
		.Select(i => i switch {
			< 0 => new PleaseMakeInference() with { Direction = Move.Left },
			> 0 => new PleaseMakeInference() with { Direction = Move.Right },
			_ => new PleaseMakeInference() with { Direction = Move.Nowhere },
		});
		/*
		*************************************************
		Desired code
		*************************************************
		.Select(i => i switch {
			< 0 => new() with { Direction = Move.Left },
			> 0 => new() with { Direction = Move.Right },
			_ => new() with { Direction = Move.Nowhere },
		});
		*************************************************
		*/
	}
	record PleaseMakeInference {
		public Move Direction { get; init; }
	}
	enum Move : sbyte {
		Left = -1,
		Nowhere = 0,
		Right = +1,
	}
}
```

**Comment by PathogenDavid** [comment]
> ```csharp
> // No inference available here despite the compiler knowing 
> // that it's explicitly dealing IEnumerable<PleaseMakeInference>
> ```

The compiler actually doesn't know that. (Well, the compiler might but the language doesn't.) The return type of `Select` is determined by the lambada rather than `directions`. So for your exact scenario you'd also need generic return-type target-typing. (See https://github.com/dotnet/csharplang/discussions/92)

----------

That being said, this doesn't work even without the involvement of `Select`: ([SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAN0QSwGYE8A+AAgEwAEAwgMoAWECADgDIQB2A5qQN4CwAUHwEhCAFlIA5OAHcA0nFySA9ggAmASRbZEcFgGM4ACgCUXQQIAKAGzjQ4AWQgBrOOs0Jte0pg2I3ygNx8pl4wnqQAvKTEAbyksXGxXq6+4aFQkpgwOtQmvAJ5ADykAAzhAHykLFJGpOkw2ZykACKYbjowmAosKbYKGAB0jHDYIQC+SKYC5SVh5ZWS1bX1TS1wbR1dET39AEqYbNSj47l5APplFVXGi1zLre2d3b1wfWIKktRapGOmI9ECI4JWkplKRLNYoHZHM5vG5dHAcnlCABmUhbeHNO7rG5sOAwPyeFgZfEA3IkgTaACuYFRT1IIFIUGAuHgCIEg2GKQAtABGI55V7vT4RIp8gS7fYhCIAal5ghJJL4LAgYDgUHoEA8hG5fW2FJY7RVfXICjA9EwVgQlEQaEweigNxIpFUUAAogAPeAIJUWdQZG4jL5AA===))

```csharp
int i = 2;
inferred = i switch {
    < 0 => new() with { Direction = Move.Left }, // error CS8754: There is no target type for 'new()'
    > 0 => new() with { Direction = Move.Right }, // error CS8754: There is no target type for 'new()'
    _ => new() with { Direction = Move.Nowhere }, // error CS8754: There is no target type for 'new()'
};
```

https://github.com/dotnet/csharplang/issues/2389 was the original proposal for adding target-typing to switch expressions during C# 8. It was never formally specified, but based on the proposed spec it sounds like it would not help in this situation. As such, I'm fairly certain this would require a language change and therefore this request belongs in [csharplang](https://github.com/dotnet/csharplang).

----------

As an aside: If you add `csharp` to the start of your code block, it will add syntax highlighting to it. IE:

~~~markdown
```csharp
Console.WriteLine("Hello, World!");
```
~~~

**Reply by alrz** [reply]
>That being said, this doesn't work even without the involvement 

new() and switch can be used together only if the target type could flow in. In that example `with` breaks the flow since it's not target typed.

https://github.com/dotnet/csharplang/issues/2926#issuecomment-553022619

**Reply by I-Omnibus** [reply]
Your example is much more concise and for what it's worth, the change to the final clause shows that the compiler has an idea of the target type for a new() keyword.

```csharp
int i = 2;
inferred = i switch {
	< 0 => new PleaseMakeInference() with { Direction = Move.Left },
	> 0 => new PleaseMakeInference() with { Direction = Move.Right },
	_ => /*new()*/ String.Empty, //Error CS0029 Cannot implicitly convert type 'string' to 'CSharpLang.PleaseMakeInference'
};
```
